torch/autograd/gradcheck.py

```diff
@@ -94,77 +113,100 @@ def get_stride(size):
             yield x_tensor, x_idx, d_idx
 
 
-def compute_gradient(fn, inputs, x, idx, delta, eps, is_mkldnn):
+def get_numerical_jacobian(fn, inputs, outputs=None, target=None, eps=1e-3, grad_out=1.0):
+    """Computes the numerical jacobian for a given fn and inputs. Outputs can be provided
+    to avoid one extra invocation of fn. Returns M * N jacobians where M is the number of
+    input tensors that require grad, and N is the number of output float/complex tensors.
+
+    input: input to `fn`
+    target: the Tensors wrt whom Jacobians are calculated (default=`input`)
+    grad_out: grad output value used to calculate gradients.
+
+    Note that `target` may not even be part of `input` to `fn`, so please be
+    **very careful** in this to not clone `target`.
+    """
+    jacobians: List[Tuple[torch.Tensor]] = []
+    if outputs is None:
+        outputs = _as_tuple(fn(inputs))
+    if target is None:
+        target = inputs
+    for i, inp in enumerate(iter_tensors(target, True)):
+        if inp.layout == torch._mkldnn and len(inputs) != 1:  # type: ignore # no attr _mkldnn
+            raise ValueError('gradcheck currently only supports functions with 1 input, but got: ',
+                             len(inputs))
+        jacobians += [get_numerical_jacobian_helper(fn, inp, inputs, outputs, eps, grad_out)]
+    return jacobians
+
+
+def compute_gradient(fn, inputs, x, idx, delta, eps, layout):
     """Perturbs inputs in-place by delta as to obtain the gradient
     of each of the outputs wrt to x at idx.
     """
     # we currently assume that the norm of delta equals eps
     assert(delta == eps or delta == (eps * 1j))
 
     def fn_out():
-        if not is_mkldnn:
-            # x is a view into input and so this works
-            return fn(inputs).clone()
-        else:
+        if layout == torch._mkldnn:  # type: ignore # no attr _mkldnn
             # convert the dense tensor back to have mkldnn layout
-            return fn([x.to_mkldnn()])
+            inp = [x.to_mkldnn()]
+        elif layout == torch.sparse_coo:
+            inp = [a.clone() for a in _as_tuple(inputs)]
```

Comment:
Temporary fix to address the issue detailed here: https://github.com/pytorch/pytorch/pull/52874. How it affects us here is that if we do `out = f(input)` and input is uncoalesced, and then we call `.coalesce()` on that input, the input will be moved off the out's backward graph, so that when we  do `.grad(out, input)` later we would always get zero (when allow_unused=True). Because we call `f` again, during get_numerical_jacobian, we may run into this issue if `f` calls `coalesce`. Note that this wasn't an issue previously because we looped through the inputs one by one and always compute numerical after we compute the analytical.